### PR TITLE
Drag&drop of text and widgets

### LINKED ISSUE #9128: Drag and drop of textual content and widgets
## 📝 Provide a description of the new feature

Extracted from [#2664](https://github.com/ckeditor/ckeditor5/issues/2664). Scenario A, B, and C with the assumption that dropping is allowed only in text positions or on the widgets.

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by niegowski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Feature (clipboard): Implemented the basic content drag and drop support. Closes #9128.

Feature (clipboard): The `contentInsertion` event is fired from `ClipboardPipeline` to enable customization of content insertion (see #9128).

Other (clipboard): The paste as plain text feature was extracted to the dedicated `PastePlainText` plugin (see #9128).

Other (engine): The `mouseup` event is fired by the `MouseObserver` (see #9128).

Other (table): The `mouseup` event is no longer fired by the `MouseEventsObserver` from the ckeditor5-table package (now handled by `MouseObserver`) (see #9128).

Internal (code-block, image, link, media-embed, paste-from-office): Updated to the changes in clipboard pipeline (see #9128).

Internal (widget): The `mousedown` event is no longer stopped from the default handling to enable content dragging (see #9128).

BREAKING CHANGE (clipboard): The `inputTransformation` event is no longer fired by the `Clipboard` plugin, now the `ClipboardPipeline` plugin is responsible for firing that event (see #9128).

BREAKING CHANGE (clipboard): The `clipboardInput` and `inputTransformation` events should not be fired or stopped in the feature code. The `data.content` property should be assigned to override the default content instead. You can stop this event only if you want to completely disable pasting/dropping of some content \[TODO migration guide link\] (see #9128).

---

### Additional information

### CODE REVIEW

**Comment by FilipTokarski** [comment]
## Steps to reproduce:
1. Create a list containing a few list items
2. Select it
3. Drag and drop somewhere else

## Expected:
Whole list gets moved, nothing remains.

## Actual:
Empty list remains at the original position.

https://user-images.githubusercontent.com/34380544/110474713-4b57b180-80e0-11eb-8fe8-669a0521306c.mp4

**Comment by FilipTokarski** [comment]
## Steps to reproduce:
1. Create a list containing only one list item
2. Select it
3. Drag and drop somewhere else

## Expected:
List gets moved to new position

## Actual:
Empty list remains at the original position, dropped content is just text
  

https://user-images.githubusercontent.com/34380544/110475785-97572600-80e1-11eb-808c-7e211eaacad9.mp4

**Comment by LukaszGudel** [comment]
## Steps to reproduce:
1. Open manual test in Safari.
2. Select some text.
3. Try to drag selected text.

## Expected:
Text can be draged and droped.

## Actual:
Selection is changing.


https://user-images.githubusercontent.com/58658048/110477264-3f212380-80e3-11eb-8791-17fe153005ed.mp4


## Notes:
- my Safari version 14.0.2
- my os: Big Sur 11.1

**Comment by FilipTokarski** [comment]
## Steps to reproduce:
1. Add some text
2. Click `Align center`
3. Select it
4. Drag into empty paragraph

## Question:
If you select the text with a mouse, or by double-clicking, after DnD alignment will not be preserved. However if you select it by triple-click, the dropped content will be centered. Similar cases with selection exist when you [copy/paste content](https://github.com/ckeditor/ckeditor5/issues/8683#issuecomment-746106567) and here in DnD, it can sometimes create unexpected behaviors ( like lists issue above ). Now, is this selection thing some bug inside editor, or should we treat this as normal behavior?

**Comment by LukaszGudel** [comment]
### Steps to reproduce:
1. Open manual test in Safari.
2. Add embedded media (youtube video).
3. Try to drag it.

### Result:
Unable to drag embedded media widget.

### Notes:
- my Safari version 14.0.2
- my os: Big Sur 11.1
- embedded medias are draggable in Chrome

**Comment by FilipTokarski** [comment]
## Steps to reproduce:
1. Add image and code block
2. Try to drag image into the codeblock

## Expected:
Because images inside code blocs are not allowed, I'd expect nothing to happen ( the image shoud stay on its place )

## Actual:
The image disappears. 
 Apart from this, if the image has text alternative or captions, their text gets pasted into code block ( but I'm not 100% sure how this should behave TBH )

  

https://user-images.githubusercontent.com/34380544/110490555-bc9f6080-80f0-11eb-8e98-5edc30bf9fa8.mp4

**Comment by niegowski** [comment]
## **Steps to reproduce:**

1.  Drag some image from local disc
2.  Drop it in the middle of some text

## **Expected:**

An image should split the paragraph exactly where it was dropped.

## **Actual:**

`findOptimalInsertionPosition()` corrects drop position to avoid splitting paragraphs.

**Comment by Reinmar** [comment]
> 1.  Drag some image from local disc
> 2.  Drop it in the middle of some text

[https://github.com/ckeditor/ckeditor5/pull/9129#issuecomment-794098198](https://github.com/ckeditor/ckeditor5/pull/9129#issuecomment-794098198) -> it's currently being changed in the inline images task (cc @oleq – please consider this case too).

---

> 1.  Add image and code block
> 2.  Try to drag image into the codeblock

[https://github.com/ckeditor/ckeditor5/pull/9129#issuecomment-794029726](https://github.com/ckeditor/ckeditor5/pull/9129#issuecomment-794029726) -> edge case, followup, but low priority.

---

> 1.  Open manual test in Safari.
> 2.  Add embedded media (youtube video).
> 3.  Try to drag it.

[https://github.com/ckeditor/ckeditor5/pull/9129#issuecomment-793967533](https://github.com/ckeditor/ckeditor5/pull/9129#issuecomment-793967533) -> followup, medium priority; it'd be at least good to know why it fails.

---

> 1.  Add some text
> 2.  Click `Align center`
> 3.  Select it
> 4.  Drag into empty paragraph

Plus:

> 1.  Create a list containing only one list item
> 2.  Select it ( or double-click on list item )
> 3.  Drag and drop somewhere else

Plus:

> 1.  Create a list containing a few list items
> 2.  Select it
> 3.  Drag and drop somewhere else

[https://github.com/ckeditor/ckeditor5/pull/9129#issuecomment-793928495,](https://github.com/ckeditor/ckeditor5/pull/9129#issuecomment-793928495,) [https://github.com/ckeditor/ckeditor5/pull/9129#issuecomment-793883321,](https://github.com/ckeditor/ckeditor5/pull/9129#issuecomment-793883321,) https://github.com/ckeditor/ckeditor5/pull/9129#issuecomment-793880326 -> these issues deserve a general followup that would include copy-paste and drag and drop. Upper-medium priority, but it's also important to have a UX analysis of this. We shouldn't make a decision in a hurry.

---

To sum up, looking at the issues reported so far, there are no critical ones that would need to be solved in this PR.

File: packages/ckeditor5-clipboard/src/clipboardpipeline.js [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+import EventInfo from '@ckeditor/ckeditor5-utils/src/eventinfo';
+
+import ClipboardObserver from './clipboardobserver';
+
+import plainTextToHtml from './utils/plaintexttohtml';
+import normalizeClipboardHtml from './utils/normalizeclipboarddata';
+import viewToPlainText from './utils/viewtoplaintext.js';
+
+// Output pipeline events overview:
```
Comment (Reinmar) [inline packages/ckeditor5-clipboard/src/clipboardpipeline.js:19]:
```suggestion
// Input pipeline events overview:
```

File: packages/ckeditor5-clipboard/src/clipboardpipeline.js [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+//                               │  inputTransformation  │
+//                               └───────────┬───────────┘
+//                                           │
+//                                ┌──────────V──────────┐
+//                                │  ClipboardPipeline  │   Calls model.insertContent().
+//                                │   contentInsertion  │
+//                                └─────────────────────┘
+//
+//
+// Input pipeline events overview:
```
Comment (Reinmar) [inline packages/ckeditor5-clipboard/src/clipboardpipeline.js:44]:
```suggestion
// Output pipeline events overview:
```

File: packages/ckeditor5-clipboard/src/dragdrop.js [hunk 3]
```diff
@@ -0,0 +1,825 @@
...
+//        │   │   ┌─────────V────────┐
+//        └───┐   │       drop       │   ( Default of clipboard pipeline ).
+//            │   └─────────┬────────┘
+//            │             │
+//            │   ┌─────────V────────┐   Resolves final data.targetRanges.
+//            │   │  clipboardInput  │   Aborts if dropping on dragged content.
+//            │   └─────────┬────────┘
+//            │             │
+//            │   ┌─────────V────────┐
+//            │   │  clipboardInput  │   ( Default of clipboard pipeline ).
```
Comment (Reinmar) [inline packages/ckeditor5-clipboard/src/dragdrop.js:79]:
```suggestion
//            │   │  clipboardInput  │   ( Default handler of the clipboard pipeline ).
```

**Comment by Mgsy** [comment]
(all browsers) Manual test issue

### Steps to reproduce

1.  Go to http://localhost:8125/ckeditor5-clipboard/tests/manual/dragdrop.html.
2.  Click somewhere outside of the editor to blur it.
3.  Ctrl + A.
4.  Start dragging from any contact item from the list and drag it to the right.

### Current result

The error occurs.

### Error

```
dragdrop.js:249 Uncaught TypeError: event.target.closest is not a function
    at HTMLUListElement.<anonymous> (dragdrop.js:249)
```

File: packages/ckeditor5-clipboard/src/dragdrop.js [hunk 4]
```diff
@@ -0,0 +1,825 @@
...
+//            │   └─────────┬────────┘
+//            │             │
+//            │ ┌───────────V───────────┐
+//            │ │  inputTransformation  │   ( Default of clipboard pipeline ).
```
Comment (niegowski) [inline packages/ckeditor5-clipboard/src/dragdrop.js:83]:
```suggestion
//            │ │  inputTransformation  │   ( Default handler of the clipboard pipeline ).
```

File: packages/ckeditor5-clipboard/src/dragdrop.js [hunk 5]
```diff
@@ -0,0 +1,825 @@
...
+//            │ └───────────┬───────────┘
+//            │             │
+//            │  ┌──────────V──────────┐
+//            │  │   contentInsertion  │   Updates the document selection to drop range.
+//            │  └──────────┬──────────┘
+//            │             │
+//            │  ┌──────────V──────────┐
+//            │  │   contentInsertion  │   ( Default of clipboard pipeline ).
```
Comment (niegowski) [inline packages/ckeditor5-clipboard/src/dragdrop.js:91]:
```suggestion
//            │  │   contentInsertion  │   ( Default handler of the clipboard pipeline ).
```

**Comment by Mgsy** [comment]
(all browsers)

### Steps to reproduce

1.  Add a link to an inline widget.
2.  Drag and drop this widget.

### Current result

The editor crashes.

### Error

```
dragdrop.js:175 Uncaught TypeError: Cannot read property 'getChildren' of undefined
    at getAndConsumeText (dragdrop.js:175)
    at getAndConsumeHCardViewElement (dragdrop.js:168)
    at model (dragdrop.js:79)
    at getModelElement (upcasthelpers.js:792)
    at UpcastDispatcher.<anonymous> (upcasthelpers.js:768)
    at UpcastDispatcher.fire (emittermixin.js:194)
    at UpcastDispatcher._convertItem (upcastdispatcher.js:249)
    at UpcastDispatcher._convertChildren (upcastdispatcher.js:282)
    at UpcastDispatcher.<anonymous> (upcasthelpers.js:896)
    at UpcastDispatcher.fire (emittermixin.js:194)
```

**Comment by LukaszGudel** [comment]
### Steps to reproduce:

1. Open manual test in Safari.
2. Add contact widget to the end of block element, as close as possible to the right side of editor.
3. Drag some text after the widget.

### Result:
Jumping widget :)

https://user-images.githubusercontent.com/58658048/110629115-1c0a7880-81a4-11eb-8cb1-f2be4e66c327.mp4

### Notes:
- low framerate on above video is not showing correct frequency of the jumping, it seems to jump much more frequent
- only in Safari
- only in Safari placing a caret after the widget in moving whole widget to the next line - in other browsers widget stays in the same line

**Comment by FilipTokarski** [comment]
## Steps to reproduce:
1. Open three editors - one in Chrome, one in Firefox and one in Safari
2. Try dragging image from one browser into another and vice versa ( for simplicity, drag them into empty paragraphs )

## Expected:
You can drag the image in both ways in all browser combinations

## Actual:
1. Chrome => Firefox - image's text alternative and caption gets pasted as text
2. Chrome => Safari - nothing happens
3. Firefox => Chrome - works fine
4. Firefox => Safari - image disappears from FF but nothing happens inf Safari
5. Safari => Chrome - image gets pasted into Chrome, but remains in Safari
6. Safari => Firefox - image gets pasted into FF, but remains in Safari

**Comment by niegowski** [comment]
> ### Steps to reproduce
> 
> 1.  Go to http://localhost:8125/ckeditor5-clipboard/tests/manual/dragdrop.html.
> 2.  Click somewhere outside of the editor to blur it.
> 3.  Ctrl + A.
> 4.  Start dragging from any contact item from the list and drag it to the right.
> 
> ### Current result
> 
> The error occurs.
> 
> ### Error
> 
> ```
> dragdrop.js:249 Uncaught TypeError: event.target.closest is not a function
>     at HTMLUListElement.<anonymous> (dragdrop.js:249)
> ```

Fixed in b7b19f6

**Comment by LukaszGudel** [comment]
Seems similar to [#issuecomment-795349643](https://github.com/ckeditor/ckeditor5/pull/9129#issuecomment-795349643)

### Steps to reproduce:
1. Open 2 manual tests in 2 Safari windows
2. Drag any content from one editor to the other one.

### Result:
Content is correctly added to second editor but stays in the first one.

**Comment by FilipTokarski** [comment]
## Steps to reproduce:
1. Create MS Word document with a list
2. Select the list and drag it into the editor

## Expected:
List is preserved

## Actual:
List items are converted into `<p>` tags

## Other info:
In the same test, pasting a list from Word works fine

## Screencast
0_dnd1

**Comment by FilipTokarski** [comment]
## Steps to reproduce:
1. Create MS Word document with some image
2. Click on the image to select it
3. Copy and paste into the editor

## Result:
```js
imageuploadediting.js:114 Uncaught TypeError: Cannot read property 'map' of null
    at Object.<anonymous> (imageuploadediting.js:114)
    at Object.fire (emittermixin.js:194)
    at fireListenerFor (bubblingemittermixin.js:162)
    at Document.fire (bubblingemittermixin.js:92)
    at Object.<anonymous> (clipboardobserver.js:57)
    at Object.fire (emittermixin.js:194)
    at fireListenerFor (bubblingemittermixin.js:162)
    at Document.fire (bubblingemittermixin.js:92)
    at ClipboardObserver.fire (domeventobserver.js:96)
    at ClipboardObserver.onDomEvent (clipboardobserver.js:83)
```

## Other info:
- on `master` and live docs/demo image gets pasted without any errors

**Comment by niegowski** [comment]
> ### Steps to reproduce
> 
> 1.  Go to http://localhost:8125/ckeditor5-clipboard/tests/manual/dragdrop.html.
> 2.  Add a link to an inline widget.
> 3.  Drag and drop this widget.
> 
> ### Current result
> 
> The editor crashes.
> 
> ### Error
> 
> ```
> dragdrop.js:175 Uncaught TypeError: Cannot read property 'getChildren' of undefined
>     at getAndConsumeText (dragdrop.js:175)
>     at getAndConsumeHCardViewElement (dragdrop.js:168)
>     at model (dragdrop.js:79)
>     at getModelElement (upcasthelpers.js:792)
>     at UpcastDispatcher.<anonymous> (upcasthelpers.js:768)
>     at UpcastDispatcher.fire (emittermixin.js:194)
>     at UpcastDispatcher._convertItem (upcastdispatcher.js:249)
>     at UpcastDispatcher._convertChildren (upcastdispatcher.js:282)
>     at UpcastDispatcher.<anonymous> (upcasthelpers.js:896)
>     at UpcastDispatcher.fire (emittermixin.js:194)
> ```

This should be resolved in 94896a9

**Comment by niegowski** [comment]
> ## Steps to reproduce:
> 
> 1.  Create MS Word document with a list
> 2.  Select the list and drag it into the editor
> 
> ## Expected:
> 
> List is preserved
> 
> ## Actual:
> 
> List items are converted into `<p>` tags
> 
> ## Other info:
> 
> In the same test, pasting a list from Word works fine

I can not reproduce this issue, checked in Chrome, dragdrop manual test, dragging from the paste from Word sample document

**Comment by LukaszGudel** [comment]
### Steps to reproduce:
1. Drag any content from Chrome or Word to Safari.

### Result:
Nothing happens. No content is added to the editor.

https://user-images.githubusercontent.com/58658048/110751542-47dd3b00-8244-11eb-9a9a-df995724f5b9.mp4

### Notes:
- DnD to Safari works from Firefox, other Safari window, or TextEdit
- my Safari version: 14.02
- my Word version: 16.44
- my OS: Big Sur 11.1

**Comment by oleq** [comment]
## Steps to reproduce

1.  Visit [http://fake.ckeditor.com:8125/ckeditor5-clipboard/tests/manual/dragdrop.html.](http://fake.ckeditor.com:8125/ckeditor5-clipboard/tests/manual/dragdrop.html.)
2.  Drag a character card into the code block

## Expected

(most likely) nothing happens because cards cannot exist as code block content

## Actual

An error is thrown

```
utils.js:99 Uncaught TypeError: Cannot read property 'match' of undefined
    at getLeadingWhiteSpaces (utils.js:99)
    at getLastOutdentableSequenceRange (outdentcodeblockcommand.js:137)
    at outdentcodeblockcommand.js:108
    at Array.some (<anonymous>)
    at OutdentCodeBlockCommand._checkEnabled (outdentcodeblockcommand.js:107)
    at OutdentCodeBlockCommand.refresh (outdentcodeblockcommand.js:41)
    at Document.<anonymous> (command.js:111)
    at Document.fire (emittermixin.js:194)
    at Document._handleChangeBlock (document.js:309)
    at Model._runPendingChanges (model.js:833)
```

**Comment by oleq** [comment]
The http://fake.ckeditor.com:8125/ckeditor5-clipboard/tests/manual/dropping.html manual test no longer makes sense (or at least its description starting with "There's no real drag&drop support yet.").

**Review by oleq — CHANGES_REQUESTED** [review]
The feature is excellent. 

I left some comments, also:

* The http://fake.ckeditor.com:8125/ckeditor5-clipboard/tests/manual/dropping.html manual test no longer makes sense (or at least its description starting with "There's no real drag&drop support yet.").

File: packages/ckeditor5-clipboard/src/clipboardpipeline.js [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+//                                 ┌─────────V────────┐
+//                                 │   view.Document  │   Processes view.DocumentFragment to text/html and text/plain
+//                                 │  clipboardOutput │   and stores results in data.dataTransfer.
+//                                 └──────────────────┘
+//
+
+/**
+ * The clipboard feature. It is responsible for intercepting the `paste` and `drop` events and
+ * passing the pasted content through the clipboard pipeline in order to insert it into the editor's content.
+ * It also handles the `cut` and `copy` events to fill the native clipboard with serialized editor's data.
```
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-clipboard/src/clipboardpipeline.js:62]: hmm why it's outdated?

**Comment by oleq** [comment]
The [http://fake.ckeditor.com:8125/ckeditor5-clipboard/tests/manual/dropping.html](http://fake.ckeditor.com:8125/ckeditor5-clipboard/tests/manual/dropping.html) manual test also stops/preventDefaults the drop event which makes the drop marker remain after the drop

**Comment by oleq** [comment]
I'd also make the triangle on top of the marker flatter:

## Current



## Proposed

Using `border-width: 8px 6px 0 6px;`



WDYT @niegowski?

**Comment by niegowski** [comment]
@ckeditor/qa-team could you please extract found issues (those that are still valid) to separate issues?
